### PR TITLE
Custom TypedShape to reference dyn Shape instead

### PR DESCRIPTION
**Pull request by Neo-Zhixing**
The existing `TypedShape::Custom` contains a user-defined shape with a type identified by a number.

This number has questionable utility. Let's make it so that `TypedShape` references the custom shape with a `&dyn Shape` instead. The user may check for the actual custom shape type using downcast.

Please let me know if there are any reasons to keep this as a `u32` I might not be aware of.

### CODE REVIEW

File: src/shape/shape.rs [hunk 1]
```diff
@@ -150,8 +150,55 @@ pub enum TypedShape<'a> {
     #[cfg(feature = "dim2")]
     #[cfg(feature = "std")]
     RoundConvexPolygon(&'a RoundConvexPolygon),
-    /// A custom user-defined shape with a type identified by a number.
-    Custom(u32),
+    /// A custom user-defined shape.
+    Custom(&'a dyn Shape),
```
Comment (Vrixyz) [inline src/shape/shape.rs:154]: Can a `dyn Shape` serialize/deserialize  ?

File: src/query/contact_manifolds/contact_manifolds_workspace.rs [hunk 2]
```diff
@@ -65,7 +65,7 @@ impl DeserializableWorkspaceData {
             DeserializableWorkspaceData::CompositeShapeShapeContactManifoldsWorkspace(w) => {
                 Some(ContactManifoldsWorkspace(Box::new(w)))
             }
-            DeserializableWorkspaceData::Custom(_) => None,
+            DeserializableWorkspaceData::Custom => None,
```
Comment (Vrixyz) [inline src/query/contact_manifolds/contact_manifolds_workspace.rs:68]: Do I understand that custom shapes are (currently) not supported by serialization/deserialization ? (at least here specifically)

Comment (Neo-Zhixing) [reply src/query/contact_manifolds/contact_manifolds_workspace.rs:68]: That is correct. And this is the same behavior as before.

**Review by sebcrozet — APPROVED** [review]
LGTM, thank you! The `u32` was supposed to be a unique id identifying the custom shape type so that some custom serialization/deserialization code could handle it. Though that custom ser/deser isn’t implemented yet and will probably not be implemented for some time.
So I think it’s is fine to remove the `u32` for now. Eventually, it will probably be re-added but as a `Uuid` instead.

File: src/shape/shape.rs (hunk 1 above)
Comment (sebcrozet) [reply src/shape/shape.rs:155]: @Vrixyz This is difficult due to the `Deserialize` trait bound having a lifetime. This is generally worked around using `erased_serde` but some additional steps are needed to make deserialization actually work. This sounds out of the scope of this PR.

**Pull request merged**